### PR TITLE
[clang-tidy] std::move improvements

### PR DESCRIPTION
**Pull request by neheb**

### CODE REVIEW

File: folly/compression/Compression.cpp [hunk 1]
```diff
@@ -70,7 +71,7 @@ namespace io {
 
 Codec::Codec(
     CodecType type,
-    Optional<int> level,
+    const Optional<int>& level,
```
Comment (yfeldblum) [inline folly/compression/Compression.cpp]: Passing `Optional<T>` by value should be fine when passing `T` by value is fine. Here and elsewhere.

File: folly/experimental/io/AsyncBase.cpp [hunk 2]
```diff
@@ -175,7 +175,7 @@ void AsyncBaseQueue::submit(AsyncBaseOp* op) {
   submit([op]() { return op; });
 }
 
-void AsyncBaseQueue::submit(OpFactory op) {
+void AsyncBaseQueue::submit(const OpFactory& op) {
   queue_.push_back(op);
```
Comment (yfeldblum) [inline folly/experimental/io/AsyncBase.cpp]: Perhaps this should be a move instead.

File: folly/executors/ThreadPoolExecutor.cpp [hunk 3]
```diff
@@ -356,7 +356,7 @@ size_t ThreadPoolExecutor::StoppedThreadQueue::size() {
   return queue_.size();
 }
 
-void ThreadPoolExecutor::addObserver(std::shared_ptr<Observer> o) {
+void ThreadPoolExecutor::addObserver(const std::shared_ptr<Observer>& o) {
```
Comment (yfeldblum) [inline folly/executors/ThreadPoolExecutor.cpp]: Perhaps `o` should be moved below instead.

File: folly/json_pointer.cpp [hunk 4]
```diff
@@ -49,7 +49,7 @@ Expected<json_pointer, json_pointer::parse_error> json_pointer::try_parse(
 json_pointer json_pointer::parse(StringPiece const str) {
   auto res = try_parse(str);
   if (res.hasValue()) {
-    return std::move(res.value());
+    return res.value();
```
Comment (yfeldblum) [inline folly/json_pointer.cpp]: Presumably we want to avoid copying `vector<string>`.

File: folly/logging/LogCategoryConfig.cpp [hunk 5]
```diff
@@ -24,7 +24,7 @@ LogCategoryConfig::LogCategoryConfig(LogLevel l, bool inherit)
 LogCategoryConfig::LogCategoryConfig(
     LogLevel l,
     bool inherit,
-    std::vector<std::string> h)
+    const std::vector<std::string>& h)
```
Comment (yfeldblum) [inline folly/logging/LogCategoryConfig.cpp]: Presumably `h` should be moved instead.

File: folly/system/MemoryMapping.cpp [hunk 6]
```diff
@@ -245,7 +245,7 @@ off_t memOpChunkSize(off_t length, off_t pageSize) {
  * - failure: false + amountSucceeded == nr bytes on which op succeeded.
  */
 bool memOpInChunks(
-    std::function<int(void*, size_t)> op,
+    const std::function<int(void*, size_t)>& op,
```
Comment (yfeldblum) [inline folly/system/MemoryMapping.cpp]: Presumably this should be templated by the op instead or should take `FunctionRef` instead. (Both by value.)

**Comment by neheb** [comment]
Replaced second commit with a different one based on feedback.

File: folly/json_pointer.cpp (hunk 4 above)
Comment (yfeldblum) [reply folly/json_pointer.cpp]: Is this from clang-tidy or is this manual?

**Comment by yfeldblum** [comment]
Note that our tooling for working with PRs actually squashes all the commits in a PR into a single commit. If you want two separate commits, can you use two separate PRs?

File: folly/json_pointer.cpp (hunk 4 above)
Comment (neheb) [reply folly/json_pointer.cpp]: clang-tidy. Maybe because of the earlier std::move.

**Comment by neheb** [comment]
OK

File: folly/json_pointer.cpp (hunk 4 above)
Comment (yfeldblum) [reply folly/json_pointer.cpp]: This seems wrong though. `res.value()` is an lvalue-ref and so this code without `std::move` logically invokes the copy-constructor of `json_pointer`, which contains a `vector<string>`. The goal of `std::move` is to invoke the move-constructor instead.

Comment (neheb) [reply folly/json_pointer.cpp]: Reverted since this is cosmetic anyway.

**Pull request closed without merging**